Implementation of AsyncListClusters functions

rsimion opened the pull request:
Fixes #1526

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1824)
<!-- Reviewable:end -->

rsimion commented:
something was cancelled. did I commit something that I was not supposed to?


On Fri, Jan 11, 2019 at 3:37 PM coryan <notifications@github.com> wrote:

> *@coryan* requested changes on this pull request.
>
> Missed a few I think, also I missed some changes in the test.
>
> Reviewed 4 of 4 files at r2.
> *Reviewable
> <https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1824#-:-LVycVkG1wPMJ69zGIYx:b-wdz3fm>*
> status: all files reviewed, 6 unresolved discussions (waiting on @rsimion
> <https://github.com/rsimion>)
>
> ------------------------------
>
> *google/cloud/bigtable/instance_admin.h, line 242 at r1
> <https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1824#-LVyKlxED-Hdwz_QA9vo:-LVyc47C493VfsoZggOs:b-egijmi>
> (raw file
> <https://github.com/googlecloudplatform/google-cloud-cpp/blob/0dac07ae13c02f7e40512020dd91c49b538f5786/google/cloud/bigtable/instance_admin.h#L242>):*
> *Previously, coryan (Carlos O'Ryan) wrote…*
>
> Indent the continued text by 4 spaces.
>
> I still do not see the indentation for @return, I meant this:
>
>  /*
> .....
>   * @return .....
>   *     currently available for querying ...
>   *     list of ...
>   *
>  .....
>   */
>
> ------------------------------
>
> *google/cloud/bigtable/tests/instance_admin_async_future_integration_test.cc,
> line 193 at r2
> <https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1824#-LVyajUq8xjMQF-RTFXu:-LVyajUq8xjMQF-RTFXv:bmaohsp>
> (raw file
> <https://github.com/googlecloudplatform/google-cloud-cpp/blob/b6a7ba41d59b5192b859d2b28627d88c5f0e02b0/google/cloud/bigtable/tests/instance_admin_async_future_integration_test.cc#L193>):*
>
>   // create cluster
>
>   auto clusters_before = instance_admin_->ListClusters(id);
>
> I do not think we need to call the synchronous version in this test
> anymore, please remove.
> ------------------------------
>
> *google/cloud/bigtable/tests/instance_admin_async_future_integration_test.cc,
> line 213 at r2
> <https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1824#-LVyaG887tDmZXEjATOU:-LVyaG89DQsV9RBhE2WX:b-6gel1k>
> (raw file
> <https://github.com/googlecloudplatform/google-cloud-cpp/blob/b6a7ba41d59b5192b859d2b28627d88c5f0e02b0/google/cloud/bigtable/tests/instance_admin_async_future_integration_test.cc#L213>):*
>
>       << " This is unexpected, as the cluster ids are"
>
>       << " generated at random.";
>
>   auto clusters_after = instance_admin_->ListClusters(id);
>
> I do not think we need to call the synchronous version anymore, please
> remove.
> ------------------------------
>
> *google/cloud/bigtable/tests/instance_admin_async_future_integration_test.cc,
> line 245 at r2
> <https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1824#-LVybxlIEwl4j-b2d6xM:-LVybxlIEwl4j-b2d6xN:b-knnvr6>
> (raw file
> <https://github.com/googlecloudplatform/google-cloud-cpp/blob/b6a7ba41d59b5192b859d2b28627d88c5f0e02b0/google/cloud/bigtable/tests/instance_admin_async_future_integration_test.cc#L245>):*
>
>   auto clusters_list_after_delete =
>
>       instance_admin_->AsyncListClusters(cq, id).get();
>
>   auto clusters_after_delete = instance_admin_->ListClusters(id);
>
> I do not think we need to call the synchronous version anymore, please
> remove and use the results from the asynchronous call.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-cpp/pull/1824#pullrequestreview-191856662>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ArwWzfy6ys_oSjGGEqyi2a2grgKDxvgSks5vCPYKgaJpZM4Z56xI>
> .
>


-- 
Robert P. Simion

rsimion commented:
Here is the output after running instance_admin_async_integration_test:

InstanceAdminAsyncFutureIntegrationTest.CreateListGetDeleteClusterTest

id: it-8haqbpwp
cluster_id_str it-8haqbpwp-cl2
clusters_list_before.clusters
projects/grass-clump-479/instances/it-8haqbpwp/clusters/it-8haqbpwp-c1

Cluster Name 
projects/grass-clump-479/instances/it-8haqbpwp/clusters/it-8haqbpwp-cl2

clusters_list_after.clusters
projects/grass-clump-479/instances/it-8haqbpwp/clusters/it-8haqbpwp-c1
projects/grass-clump-479/instances/it-8haqbpwp/clusters/it-8haqbpwp-cl2

clusters_list_after_delete.clusters
projects/grass-clump-479/instances/it-8haqbpwp/clusters/it-8haqbpwp-c1


//------------------------------------------------------------------------

InstanceAdminAsyncFutureIntegrationTest.AsyncListAllClustersTest

instance1 
projects/grass-clump-479/instances/it-pl42hgjq
instance2 
projects/grass-clump-479/instances/it-27lm3k70
clusters_list.clusters before
Size: 19
projects/grass-clump-479/instances/cpp-instance/clusters/cpp-instance-c1
projects/grass-clump-479/instances/cpp-integration-tests/clusters/cpp-integration-tests-c1
projects/grass-clump-479/instances/dotnet-perf-2/clusters/dotnet-perf-2-cluster
projects/grass-clump-479/instances/endurance/clusters/endurance-c1
projects/grass-clump-479/instances/it-1ohkpa34/clusters/it-1ohkpa34-c1
projects/grass-clump-479/instances/it-27lm3k70/clusters/it-27lm3k70-c1
projects/grass-clump-479/instances/it-2w2rjzrw/clusters/it-2w2rjzrw-c1
projects/grass-clump-479/instances/it-98608odh/clusters/it-98608odh-c1
projects/grass-clump-479/instances/it-jsdv1tdd/clusters/it-jsdv1tdd-c1
projects/grass-clump-479/instances/it-krhzseyc/clusters/it-krhzseyc-c1
projects/grass-clump-479/instances/it-pl42hgjq/clusters/it-pl42hgjq-c1
projects/grass-clump-479/instances/it-r8ozr8i4/clusters/it-r8ozr8i4-c1
projects/grass-clump-479/instances/it-rypktfc4/clusters/it-rypktfc4-c1
projects/grass-clump-479/instances/it-ypphdg4i/clusters/it-ypphdg4i-c1
projects/grass-clump-479/instances/java-samples/clusters/java-samples-c1
projects/grass-clump-479/instances/node-client-performance/clusters/node-client-performance-c1
projects/grass-clump-479/instances/python-write/clusters/python-write-c1
projects/grass-clump-479/instances/shared-perf/clusters/shared-perf-cluster
projects/grass-clump-479/instances/shared-perf-2/clusters/shared-perf-2-cluster

So far it looks like it works.

sduskis commented on google/cloud/bigtable/tests/instance_admin_async_future_integration_test.cc:
```diff
@@ -177,21 +187,37 @@ TEST_F(InstanceAdminAsyncFutureIntegrationTest,
+      << "The Cloud Bigtable service (or emulator) reports that it could not"
+      << " retrieve the information for some locations. This is typically due"
+      << " to an outage or some other transient condition.";
+  ASSERT_FALSE(IsClusterPresent(clusters_list_before.clusters, cluster_id_str))
```
IsClusterPresent takes a cluster name, not a cluster id.  The original code was wrong.  Please use `cluster.name()` instead of id.

rsimion replied:
At that line (197) the cluster variable has not been created yet.
It is created at line 206.
So I added these  extra-checks at line 219 and 220:
  EXPECT_FALSE(IsClusterPresent(clusters_list_before.clusters, cluster.name()));
  EXPECT_TRUE(IsClusterPresent(clusters_list_after.clusters, cluster.name()));

rsimion commented:
Made all the requested changes.

coryan commented:
Sorry, I meant to add.  I think you addressed most issues, there is just a couple of assertions that could pass even when not true, please fix.

rsimion commented:
Thanks.

file instance_admin_async_Integration_test.cc

I modified the function:

bool IsInstanceNamePresentInClusterList(
    std::vector<**btadmin::Cluster**> const& clusters,
    std::string const& instance_name) {
  return clusters.end() !=
         std::find_if(clusters.begin(), clusters.end(),
                      [&instance_name](btadmin::Cluster const& i) {
                        return i.name().find(instance_name)  != std::string::npos;
                      });
}

I replaced 
ASSERT_GT(cluster.name().size(), 0);
with
ASSERT_FALSE(cluster.name().empty());

I ran the emulator and everything worked fine.

I removed 
ASSERT_FALSE(IsClusterPresent(clusters_list_before.clusters, cluster_id_str))
ASSERT_FALSE(IsClusterPresent(clusters_list_after.clusters, cluster_id_str))

Did I miss something?
